src/core/MOM_porous_barriers.F90

```diff
@@ -0,0 +1,168 @@
+                                                                    !! thermodynamic variables.
+  real, dimension(SZI_(G),SZJ_(G),SZK_(G)+1), intent(out) :: eta    !< layer interface heights
+                                                                    !! [Z ~> m] or 1/eta_to_m m).
+  real, dimension(SZI_(G),SZJ_(G)), optional, intent(in)  :: eta_bt !< optional barotropic
+             !! variable that gives the "correct" free surface height (Boussinesq) or total water
+             !! column mass per unit area (non-Boussinesq).  This is used to dilate the layer.
+             !! thicknesses when calculating interfaceheights [H ~> m or kg m-2].
+  integer,                          optional, intent(in)  :: halo_size !< width of halo points on
+                                                                       !! which to calculate eta.
+
+  real,                             optional, intent(in)  :: eta_to_m  !< The conversion factor from
+             !! the units of eta to m; by default this is US%Z_to_m.
+  type(porous_barrier_ptrs),           intent(inout) :: pbv  !< porous barrier fractional cell metrics
+
+  !local variables
+  integer ii, i, j, k, nk, isd, ied, jsd, jed, IsdB, IedB, JsdB, JedB
+  real w_layer, & ! fractional open width of layer interface [nondim]
+       A_layer, & ! integral of fractional open width from bottom to current layer[Z ~> m]
+       A_layer_prev, & ! integral of fractional open width from bottom to previous layer [Z ~> m]
+       eta_s, & ! layer height used for fit [Z ~> m]
+       eta_prev ! interface height of previous layer [Z ~> m]
+  isd = G%isd; ied = G%ied; jsd = G%jsd; jed = G%jed
+  IsdB = G%IsdB; IedB = G%IedB; JsdB = G%JsdB; JedB = G%JedB
+
+  !eta is zero at surface and decreases downward
+
+  nk = SZK_(G)
+
+  !currently no treatment for using optional find_eta arguments if present
+  call find_eta(h, tv, G, GV, US, eta)
+
+  do j=jsd,jed; do I=IsdB,IedB
+    if (G%porous_DavgU(I,j) < 0.) then
+      do K = nk+1,1,-1
+        eta_s = max(eta(I,j,K), eta(I+1,j,K)) !take shallower layer height
+        if (eta_s <= G%porous_DminU(I,j)) then
+          pbv%por_layer_widthU(I,j,K) = 0.0
+          A_layer_prev = 0.0
+          if (K < nk+1) then
+            pbv%por_face_areaU(I,j,k) = 0.0; endif
+        else
+          call calc_por_layer(G%porous_DminU(I,j), G%porous_DmaxU(I,j), &
+            G%porous_DavgU(I,j), eta_s, w_layer, A_layer)
+          pbv%por_layer_widthU(I,j,K) = w_layer
+          if (k <= nk) then
+            if ((eta_s - eta_prev) > 0.0) then
+              pbv%por_face_areaU(I,j,k) = (A_layer - A_layer_prev)/&
+                   (eta_s-eta_prev)
+            else
+              pbv%por_face_areaU(I,j,k) = 0.0; endif
+          endif
+          eta_prev = eta_s
+          A_layer_prev = A_layer
+        endif
+      enddo
+    endif
+  enddo; enddo
+
+  do i=isd,ied; do J=JsdB,JedB
```

Comment:
The order of these loops also need to be inverted, to `do J=JsdB,JedB ; do i=isd,ied`, so that we are taking a stride of 1 through the memory.